_posts/2018-05-04-design-of-task-graph-generation.md

```diff
@@ -0,0 +1,253 @@
+---
+title: 'Design of Task-Graph Generation'
+layout: post
+date:   2018-05-04 15:00:00
+categories: [mozilla,taskcluster]
+---
+
+Almost two years ago, [Bug 1258497](https://bugzilla.mozilla.org/show_bug.cgi?id=1258497)
+introduced a new system for generating the graph of tasks required for each
+push to a Firefox source-code repository. Work continues to modify the expected
+tasks and add features, but the core design is stable. Lots of Firefox
+developers have encountered this system as they add or modify a job or try to
+debug why a particular task is failing. So this is a good time to review the
+system design at a high level.
+
+A quick note before beginning: the task-graph generation system is implemented
+entirely in the Firefox source tree, and is administered as a sub-module of the
+Build Config module. While it is designed to interface with Taskcluster, and
+some of the authors are members of the Taskcluster team, it is not a part of
+Taskcluster itself.
+
+# Requirements
+
+A task is a unit of work in the aptly-named Taskcluster platform. This might be
+a Firefox build process, or a run of a chunk of a test suite. More esoteric
+tasks include builds of the toolchains and OS environments used by other tasks;
+cryptographic signing of Firefox installers; configuring Balrog, the service
+behind Firefox's automatic updates; and pushing APKs to the Google Play Store.
+
+A task graph is a collection of tasks linked by their dependencies. For
```

Comment:
nit: but defining _terms_ using italic is super neat :)

Also do we write `task graph` or `task-graph`.

Okay, if you're not aiming for a Pulitzer, maybe just ignore this comment.

Reply:
Oh, I'm going for a Pulitzer!